### PR TITLE
[Posix][Filesystem] Fixed SMB logging

### PR DESCRIPTION
**Pull request by hugbug**
## Description

The logging of SMB component (libsmbclient) requires setup of a logging function.

## Motivation and context

Kodi has a set of logging options for troubleshooting. One of the options activates logging for SMB component. However it did not work. The required logging function for libsmbclient was not properly set.

There was already a function defined but not used. It seems that function was used 15 years ago but only on Windows. Currently libsmbclient isn't used on Windows at all. I corrected the function for current libsmbclient requirements and added the code to attach the callback to libsmbclient.

Small example of a log after fix:
```
2024-01-02 22:52:24.521 T:11488    debug <general>: smb: SMBC_server: server_n=[192.168.0.105] server=[192.168.0.105]
2024-01-02 22:52:24.521 T:11488    debug <general>: smb:  -> server_n=[192.168.0.105] server=[192.168.0.105]
2024-01-02 22:52:24.523 T:11488    debug <general>: smb: Connecting to 192.168.0.105 at port 445
2024-01-02 22:52:24.527 T:11488    debug <general>: smb: socket options: SO_KEEPALIVE=0, SO_REUSEADDR=0, SO_BROADCAST=0, 
TCP_NODELAY=1, TCP_KEEPCNT=9, TCP_KEEPIDLE=7200, TCP_KEEPINTVL=75, IPTOS_LOWDELAY=0, IPTOS_THROUGHPUT=0, SO_REUSEPORT=0, SO_SNDBUF=524288,
SO_RCVBUF=1048576, SO_SNDLOWAT=1, SO_RCVLOWAT=1, SO_SNDTIMEO=0, SO_RCVTIMEO=0, TCP_QUICKACK=1, TCP_DEFER_ACCEPT=0, TCP_USER_TIMEOUT=0
…..
2024-01-02 22:52:54.428 T:11538    debug <general>: smb: smbc_stat(smb://username:password@192.168.0.105/Testvideos/Video.mp4)
2024-01-02 22:52:54.428 T:11538    debug <general>: smb: SMBC_getatr: sending qpathinfo
2024-01-02 22:52:54.475 T:11538    debug <general>: smb: smbc_read(0xb321cbf0, 1048576)
2024-01-02 22:52:54.519 T:11538    debug <general>: smb:   --> 1048576
2024-01-02 22:52:54.520 T:11538    debug <general>: smb: smbc_read(0xb321cbf0, 1048576)
2024-01-02 22:52:54.544 T:11538    debug <general>: smb:   --> 1048576

```

## How has this been tested?

Kodi is running on Mac with Intel CPU. SMB share is on PC with Windows 11 Pro.

In Kodi under Settings -> System -> Logging the option "Enable component-specific logging" was activated and in the option "Specify component-specific logging" the item "Verbose logging for the SMB library" was selected. Kodi was restarted. In Kodi File Manager a new source pointing to a network SMB share was added. A video file located on this SMB source was played. Then kodi.log was inspected.

## What is the effect on users?

The logging of SMB component works now. The logs can be helpful for troubleshooting.

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/platform/posix/filesystem/SMBFile.cpp [hunk 1]
```diff
@@ -34,9 +34,9 @@
 
 using namespace XFILE;
 
-void xb_smbc_log(const char* msg)
+void xb_smbc_log(void* private_ptr, int level, const char* msg)
 {
-  CLog::Log(LOGINFO, "{}{}", "smb: ", msg);
+  CLog::Log(LOGDEBUG, "smb: {}", msg);
```
Comment (neo1973) [inline xbmc/platform/posix/filesystem/SMBFile.cpp:39]:
```suggestion
  const int logLevel = [level]()
  {
    switch (level)
    {
      case 0:
        return LOGWARNING;
      case 1:
        return LOGINFO;
      default:
        return LOGDEBUG;
    }
  }();
  CLog::Log(logLevel, "smb: {}", msg);
```
As `xb_smbc_log` will be called for critical errors and warnings even if component logging is disabled, it's probably a good idea to log these messages as such. I assume the `level` passed in here corresponds to the `-d|--debuglevel=DEBUGLEVEL` flag described in https://www.samba.org/samba/docs/current/man-html/smbclient.1.html.

Comment (hugbug) [reply xbmc/platform/posix/filesystem/SMBFile.cpp:39]: Nice idea. Let me test this.

**Review by neo1973 — APPROVED** [review]
Can't runtime test as I don't have a Samba share setup but code looks good to me :slightly_smiling_face:

**Review by thexai — APPROVED** [review]
Tested on Shield and works as expected.

**Comment by hugbug** [comment]
Smb logs contain personal information such as username, password, servername, it-address.

Here is an example of logged smb session: [Kodi-smb.log](https://paste.kodi.tv/ogizapacug). I filtered only smb messages and removed timeframes for easier viewing. Before posting the log I also replaced username, password and servername with strings REAL-USERNAME, REAL-PASSWORD and REAL-SERVERNAME.

Few examples:
```
smb: parsed path: fname='smb://REAL-USERNAME:REAL-PASSWORD@192.168.0.105/FireTV' server='192.168.0.105' share='FireTV' path='' options=''
smb: SMBC_check_options(): server='192.168.0.105' share='FireTV' path='' options=''
smb: SMBC_server: server_n=[192.168.0.105] server=[192.168.0.105]
smb:  -> server_n=[192.168.0.105] server=[192.168.0.105]
smb: Connecting to 192.168.0.105 at port 445
smb: cli_session_setup_spnego_send: Connect to 192.168.0.105 as REAL-USERNAME@WORKGROUP using SPNEGO
smb:                         AvNbDomainName           : 'REAL-SERVERNAME'
smb:                                 AvNbComputerName         : 'REAL-SERVERNAME'
smb:             DomainName               : 'WORKGROUP'
smb:             UserName                 : 'REAL-USERNAME'
smb: smbc_stat(smb://REAL-USERNAME:REAL-PASSWORD@192.168.0.105/FireTV/.DS_Store)
smb: SMBC_getatr: sending qpathinfo
smb: smbc_getxattr(smb://REAL-USERNAME:REAL-PASSWORD@192.168.0.105/FireTV/.DS_Store, system.dos_attr.mode)
```

Should we try to filter out the sensible info similar to `CURL:GetRedacted`?

I'm afraid we can't achieve 100% clean logs. The amount of logging in Samba is overwhelming. I've counted more than 18000 log messages in Samba source code.

What's your opinion on this?

**Comment by hugbug** [comment]
@thexai, your comment came up as I was writing mine ;)

Would it be enough to replace only username:password combination as you suggested?

**Comment by thexai** [comment]
Yes, I think this is sufficient. As proof of concept:

```c++
  std::string msg = "parsed path: fname='smb://REAL-USERNAME:REAL-PASSWORD@192.168.0.105/FireTV'";
  std::regex redact("(\\w+://)\\S+:\\S+@");
  CLog::Log(LOGINFO, "Test: {}", std::regex_replace(msg, redact, "$1USERNAME:PASSWORD@"));
```

**Comment by hugbug** [comment]
Thank you. That worked like a charm.

I use `strchr` as a pre-condition for RE to reduce the load on RE compiler/executer. I considered making RE variable global or a member of `CSMB` to avoid compilation of RE on every usage. At the end I decided against that to not overcomplicate the code. The debug logging of SMB is not active normally and the strchr-check should filter out most of the messages, especially all that messages for chunk transfers.

File: xbmc/platform/posix/filesystem/SMBFile.cpp [hunk 2]
```diff
@@ -29,14 +29,35 @@
...
+        return LOGINFO;
+      default:
+        return LOGDEBUG;
+    }
+  }();
+
+  if (strchr(msg, '@'))
+  {
+    // redact User/pass in URLs
+    std::regex redact("(\\w+://)\\S+:\\S+@");
```
Comment (neo1973) [inline xbmc/platform/posix/filesystem/SMBFile.cpp:56]:
```suggestion
    static const std::regex redact("(\\w+://)\\S+:\\S+@");
```
This should take care of a part of your concern:

> reduce the load on RE compiler/executer

Keeping the `strchr` in addition doesn't hurt :slightly_smiling_face:

File: xbmc/platform/posix/filesystem/SMBFile.cpp [hunk 3]
```diff
@@ -29,14 +29,35 @@
...
+      case 0:
+        return LOGWARNING;
+      case 1:
+        return LOGINFO;
+      default:
+        return LOGDEBUG;
+    }
+  }();
+
+  if (strchr(msg, '@'))
```
Comment (thexai) [inline xbmc/platform/posix/filesystem/SMBFile.cpp:53]:
```suggestion
  if (std::strchr(msg, '@'))
```

Comment (thexai) [reply xbmc/platform/posix/filesystem/SMBFile.cpp:53]: And include <cstring>

File: xbmc/platform/posix/filesystem/SMBFile.cpp (hunk 2 above)
Comment (hugbug) [reply xbmc/platform/posix/filesystem/SMBFile.cpp]: Nice touch. Thank you.

File: xbmc/platform/posix/filesystem/SMBFile.cpp (hunk 3 above)
Comment (hugbug) [reply xbmc/platform/posix/filesystem/SMBFile.cpp]: Thank you.

It compiles without explicit include of `<cstring>`. It must have got included through some other include. Do you still prefer an explicit include to be added?

(I don't want to pollute the history with super-small commits, unless necessary. I also don't want to squash commits to respect authors/reviewers.)

Comment (hugbug) [reply xbmc/platform/posix/filesystem/SMBFile.cpp]: I've tried `squash` locally and it kept all the authors. I'll squash the commits then for cleaner history.

Comment (thexai) [reply xbmc/platform/posix/filesystem/SMBFile.cpp]: > It compiles without explicit include of <cstring>. It must have got included through some other include. Do you still prefer an explicit include to be added?

Technically it is the most correct.

> (I don't want to pollute the history with super-small commits, unless necessary. I also don't want to squash commits to respect authors/reviewers.)

You can amend last commit

**Comment by hugbug** [comment]
I've applied all suggested changes. Thank you very much.

**Comment by neo1973** [comment]
jenkins build this please

**Comment by neo1973** [comment]
Merging, the Jenkins error is unrelated. Thanks again for the contribution :slightly_smiling_face:

**Pull request merged**